svm/src/account_loader.rs

```diff
@@ -1382,35 +1559,42 @@ mod tests {
         let mut loaded_programs = ProgramCacheForTxBatch::default();
         loaded_programs.replenish(key2.pubkey(), Arc::new(ProgramCacheEntry::default()));
 
+        let mut account_loader =
+            AccountLoader::new_with_account_cache_capacity(&mock_bank, None, loaded_programs, 1);
+
         let sanitized_transaction = SanitizedTransaction::new_for_tests(
             sanitized_message,
             vec![Signature::new_unique()],
             false,
         );
         let result = load_transaction_accounts(
-            &mock_bank,
+            &mut account_loader,
             sanitized_transaction.message(),
             LoadedTransactionAccount::default(),
             &ComputeBudgetLimits::default(),
             &mut error_metrics,
-            None,
             &FeatureSet::default(),
             &RentCollector::default(),
-            &loaded_programs,
         );
 
-        assert_eq!(result.err(), Some(TransactionError::AccountNotFound));
+        assert_eq!(result.err(), Some(TransactionError::ProgramAccountNotFound));
     }
 
-    #[test]
-    fn test_load_transaction_accounts_program_account_executable_bypass() {
-        // currently, the account loader retrieves read-only non-instruction accounts from the program cache
-        // it creates a mock AccountSharedData with the executable flag set to true
-        // however, it does not check whether these accounts are actually executable before doing so
-        // this affects consensus: a transaction that uses a cached non-executable program executes and fails
-        // but if the transaction gets the program from accounts-db, it will be dropped during account loading
-        // this test enforces the current behavior, so that future account loader changes do not break consensus
-
+    // currently, the account loader retrieves read-only non-instruction accounts from the program cache
+    // it creates a mock AccountSharedData with the executable flag set to true
+    // however, it does not check whether these accounts are actually executable before doing so
+    // this affects consensus: a transaction that uses a cached non-executable program executes and fails
+    // but if the transaction gets the program from accounts-db, it will be dropped during account loading
+    // this test enforces the current behavior, so that future account loader changes do not break consensus
+    //
+    // account cache has its own code path that accesses program cache, so we test hit and miss
+    // we also test bpf_loader and bpf_loader_upgradeable, because accounts owned by the latter can behave strangely
+    // all cases should produce the same results
+    #[test_matrix([bpf_loader::id(), bpf_loader_upgradeable::id()], [false, true])]
+    fn test_load_transaction_accounts_program_account_executable_bypass(
+        program_owner: Pubkey,
+        clear_account_cache: bool,
+    ) {
```

Comment:
also more coverage for that case here, the code is the same but the test matrix ensures we hit all codepaths and a normal loader vs the upgradeable loader. there isnt actually any difference in behavior between loaderv3 buffers and anything else tho. the program cache pulls in empty accounts owned by the loaders too